docs: new tutorial validation of messages/events (#1061)
asyncapi/website: pages/docs/tutorials/message-validation.md

---
title: Message validation in runtime
description: In this tutorial, you'll learn how to validate AsyncAPI messages (events).

> weight: 130

Reviewer: @alequetzalli Kindly look into the weight as this weight may be changing the order of the pages inside the docs, which can be not preferred by you. You need to look inside the Tree presented on the Sidebar of Docs pages carefully.

![image](https://user-images.githubusercontent.com/76521428/208171271-e0b50429-b784-4d7f-bd94-9e3e789f0593.png)

I don't think you prefer to have this tutorial in between the Streetlights tutorial - Interactive and the normal Streetlights tutorial.

Reviewer 2: @akshatnema I know you're not up-to-date on what's going on in the Docs roadmap, but your comment doesn't apply because both interactive and static streetlights tutorials will be removed when these are merged.

Thank you for reaching out and asking the question though! Always good to ask and understand what's going on right now

1. Start your generated application.

<CodeBlock language="bash">
{`npm start`}
</CodeBlock>

2. Let's send a message:

<CodeBlock language="bash">
  {`mqtt pub -t 'light/measured' -h 'test.mosquitto.org' -m '{"id": 1, "lumens": "3", "sentAt": "2017-06-07T12:34:32.000Z"}'`}
</CodeBlock>
  
Go back to the previous terminal and check if your application logged the streetlight condition you just sent, with errors related to the invalid message. You should see something displayed in the terminal similar to the following:

<CodeBlock language="bash">
  {`light/measured was received:
{ id: 1, lumens: '3', sentAt: '2017-06-07T12:34:32.000Z' }
❗  Message Rejected. data.lumens should be integer`}
</CodeBlock>
 
 <CodeBlock language="yaml" highlightedLines={[10,11]}>
  {`  message:
        name: lumensInfo
        payload:
          type: object
          properties:
            id:
              type: integer
              minimum: 0
              description: Id of the streetlight.
            lumens:
              type: integer
              minimum: 0
              description: Light intensity measured in lumens.`}
</CodeBlock>
  

> Here, you can see that the property `lumens` has type `integer`, but you are sending a message with type `string`.

Reviewer 2: @derberg I just noticed that there is a styling issue here but I don't see any issues in the markdown. Why do you think this is happening? O_o

<img width="500" alt="Screen Shot 2022-12-05 at 6 15 55 PM" src="https://user-images.githubusercontent.com/19964402/205792320-af8ffa94-9adc-4464-b4c4-c6e766fa1169.png">

Author: @alequetzalli fixed it

3. Send a correct message to your application:

<CodeBlock language="bash">
  {`mqtt pub -t 'light/measured' -h 'test.mosquitto.org' -m '{"id": 1, "lumens": 3, "sentAt": "2017-06-07T12:34:32.000Z"}'`}
</CodeBlock>

You can see that your generated application received a message in the terminal:

<CodeBlock language="bash">
  {`light/measured was received:
{ id: 1, lumens: 3, sentAt: '2017-06-07T12:34:32.000Z' }`}
</CodeBlock>

This indicates that your message is valid and the application recieved it correctly.

## Summary
In this tutorial, you learned how to connect your generated application to an MQTT broker, send messages through it, identify when an invalid message is sent to your application, and how to correct an invalid message. 

---

<DocsButton
  suggestions={[
    {

>       href: '/docs/tutorials/generate-code',

Reviewer: This particular page isn't available in the Documentation. I prefer to change this `back` button route if you want to merge this PR now or wait till the mentioned page gets added to the Docs.

      title: 'Generate code',
      type:'back',
    },
    {

>       href: '/docs/tools',

Reviewer: As far as I can see in the Docs tree presented in the sidebar, it is not aligned properly. If you want to make the next button point towards the `Tools - Overview`, kindly change the `weight` parameter accordingly.

---
title: Message validation
description: In this tutorial, you'll learn how to validate AsyncAPI messages (events).

weight: 130
---

# Introduction
In this tutorial, you'll learn how to validate messages (events) that are sent to your AsyncAPI application.

# Background context
Message validation can be performed at both the producer and consumer levels. Message validation requires the participation of the producer, consumer, and broker. We will learn how to validate messages at the consumer level by discarding invalid messages based on the parameters provided.

> You will be using [Eclipse Mosquitto](https://mosquitto.org/) broker. The MQTT protocol provides a lightweight method of carrying out messaging using a publish/subscribe model.
> You will be using MQTT client that runs an MQTT library and connects to an MQTT broker over a network. Here publishers and subscribers are MQTT clients. The publisher and subscriber labels refer to whether the client is currently publishing messages or subscribed to receive messages.

Reviewer 2: ```suggestion
You will be using the [Eclipse Mosquitto](https://mosquitto.org/) broker. The MQTT protocol provides a lightweight method of messaging using a publish/subscribe model. You will also use an MQTT client that runs an MQTT library and connects to an MQTT broker over a network. Here publishers and subscribers are MQTT clients. The publisher and subscriber labels refer to whether the client is publishing or subscribing to receive messages.

```

> In previous tutorial, you have generated your application using [AsyncAPI Generator](https://github.com/asyncapi/generator) Node.js template.
> Now you will be validating the messages/events which you will be sending to your application using Mosquitto broker and MQTT client.

Reviewer 2: ```suggestion
In the previous tutorial, you generated your application using the [AsyncAPI Generator](https://github.com/asyncapi/generator) Node.js template. Now you will be validating the messages (events) which you will be sending to your application using a Mosquitto broker and MQTT client.

```

            lumens:
              type: integer
              minimum: 0
              description: Light intensity measured in lumens.`}
  </CodeBlock>
  
Here, you can see that the property `lumens` has type `integer`, but you are sending a message with type `string`.

3. Send a correct message to your application:

<CodeBlock language="bash">
  {`mqtt pub -t 'light/measured' -h 'test.mosquitto.org' -m '{"id": 1, "lumens": 3, "sentAt": "2017-06-07T12:34:32.000Z"}'`}
  </CodeBlock>

You can see that your generated application received a message in the terminal:

<CodeBlock language="bash">
  {`light/measured was received:
{ id: 1, lumens: 3, sentAt: '2017-06-07T12:34:32.000Z' }`}
  </CodeBlock>

This indicates that your message is valid and it is recieved by the application correctly.

# Summary 
In this tutorial, you learned how to connect your generated application to an MQTT broker, send messages through it, how to identify when an invalid message is sent to your application, and how you can correct the invalid message. 

---

<DocsButton
  suggestions={[
    {
      href: '/docs/tutorials/generate-code',
      title: 'Generate code',
      type:'back',
    },
    {

>       href: '/docs/guides',

Reviewer 2: ```suggestion
      href: '/docs/tools',

```

>       title: 'Overview',

Reviewer 2: ```suggestion
      title: 'Tools - Overview',

```
😜🎉

yooo, the next content bucket AFTER tutorials is `Tools`, not Guides 

<img width="500" alt="Screen Shot 2022-11-16 at 5 08 47 PM" src="https://user-images.githubusercontent.com/19964402/202329343-2fb87a0e-fe8d-4186-8a2a-21c26411d87b.png">

---

> title: Message validation

Reviewer 3: ```suggestion
title: Message validation in runtime
```

> description: In this tutorial, you'll learn how to validate AsyncAPI messages (events).

Reviewer 3: I think we can just stick to `message` and if you want in intro just reference message concept document where both are explained, message and event

weight: 130
---

> # Introduction

Reviewer 3: ```suggestion
## Introduction
```

In this tutorial, you'll learn how to validate messages (events) that are sent to your AsyncAPI application.

> # Background context

Reviewer 3: ```suggestion
## Background context
```

Message validation can be performed at both the producer and consumer levels. Message validation requires the participation of the producer, consumer, and broker. We will learn how to validate messages at the consumer level by discarding invalid messages based on the parameters provided.

You will be using the [Eclipse Mosquitto](https://mosquitto.org/) broker. The MQTT protocol provides a lightweight method of messaging using a publish/subscribe model. You will also use an MQTT client that runs an MQTT library and connects to an MQTT broker over a network. Here publishers and subscribers are MQTT clients. The publisher and subscriber labels refer to whether the client is publishing or subscribing to receive messages.

> In the previous tutorial, you generated your application using the [AsyncAPI Generator](https://github.com/asyncapi/generator) Node.js template. Now you will be validating the messages (events) which you will be sending to your application using a Mosquitto broker and MQTT client.

Reviewer 3: since tutorials are supposed to be standalone please add instructions to:
- add instruction to quickly generate asyncapi document 
   ```
   asyncapi new --example=tutorial.yml --no-tty`
   ```
- add instruction to quickly generate server 
   ```
   asyncapi generate fromTemplate asyncapi.yaml @asyncapi/nodejs-template -o output -p server=mosquitto
   cd output && npm install
   ```

Author: done

> #  Validate messages

Reviewer 3: ```suggestion
## Validate messages
```

In this step, we will send a message to our application using an MQTT broker and check the errors logged when you accidentally send an invalid message.

1. Start your generated application.

<CodeBlock language="bash">
{`npm start`}
</CodeBlock>

2. Now let's try to send a message:

<CodeBlock language="bash">
  {`mqtt pub -t 'light/measured' -h 'test.mosquitto.org' -m '{"id": 1, "lumens": "3", "sentAt": "2017-06-07T12:34:32.000Z"}'`}
  </CodeBlock>
  

>  Go back to the previous terminal and check if your application logged the streetlight condition you just sent, with errors related to the invalid message.

Reviewer 3: ```suggestion
 Go back to the previous terminal and check if your application logged the streetlight condition you just sent, with errors related to the invalid message. You should see something like this displayed in the terminal:
```

and please add code snippet that shows example validation message

Author: OKay so I tried doing it in killer coda and it didn't show me any errors when I changed number to string but it should right?

Reviewer 3: hmmm, I just generated code locally and works good:

```bash
SUB  Subscribed to light/measured
Streetlights API Simplified 1.0.0 is ready! 

🔗  MQTT adapter is connected!
← light/measured was received:
{ id: 1, lumens: '3', sentAt: '2017-06-07T12:34:32.000Z' }
❗  Message Rejected. data.lumens should be integer
```

I recommend playing with it on local, especially as tutorial author

 
 <CodeBlock language="yaml" highlightedLines={[10,11]}>
  {`  message:
        name: lumensInfo
        payload:
          type: object
          properties:
            id:
              type: integer
              minimum: 0
              description: Id of the streetlight.
            lumens:
              type: integer
              minimum: 0
              description: Light intensity measured in lumens.`}
  </CodeBlock>
  
Here, you can see that the property `lumens` has type `integer`, but you are sending a message with type `string`.

3. Send a correct message to your application:

<CodeBlock language="bash">
  {`mqtt pub -t 'light/measured' -h 'test.mosquitto.org' -m '{"id": 1, "lumens": 3, "sentAt": "2017-06-07T12:34:32.000Z"}'`}
  </CodeBlock>

You can see that your generated application received a message in the terminal:

<CodeBlock language="bash">
  {`light/measured was received:
{ id: 1, lumens: 3, sentAt: '2017-06-07T12:34:32.000Z' }`}
  </CodeBlock>

This indicates that your message is valid and it is recieved by the application correctly.

> # Summary 

Reviewer 3: ```suggestion
## Summary
```

In this tutorial, you learned how to connect your generated application to an MQTT broker, send messages through it, how to identify when an invalid message is sent to your application, and how you can correct the invalid message. 

---

<DocsButton
  suggestions={[
    {
      href: '/docs/tutorials/generate-code',
      title: 'Generate code',
      type:'back',
    },

>     {
>       href: '/docs/tools',
> 
>       title: 'Tools - Overview',
> 
>       type:'next',
>     }

Reviewer 3: ```suggestion
    {
      href: '/docs/tools',
      title: 'Tools - Overview',
      type:'next',
    }
```

preview building is failing https://app.netlify.com/sites/asyncapi-website/deploys/6376426f0389d400099df905
not sure if because of this, but let us see

---
title: Message validation in runtime
description: In this tutorial, you'll learn how to validate AsyncAPI messages (events).

weight: 130
---

## Introduction
In this tutorial, you'll learn how to validate messages (events) that are sent to your AsyncAPI application.

## Background context
Message validation can be performed at both the producer and consumer levels. Message validation requires the participation of the producer, consumer, and broker. We will learn how to validate messages at the consumer level by discarding invalid messages based on the parameters provided.

You will be using the [Eclipse Mosquitto](https://mosquitto.org/) broker. The MQTT protocol provides a lightweight method of messaging using a publish/subscribe model. You will also use an MQTT client that runs an MQTT library and connects to an MQTT broker over a network. Here publishers and subscribers are MQTT clients. The publisher and subscriber labels refer to whether the client is publishing or subscribing to receive messages.

In the previous tutorial, you generated your application using the [AsyncAPI Generator](https://github.com/asyncapi/generator) Node.js template.
<Remember>

> If you did not follow the previous tutorial and do not have an `asyncapi.yaml` file ready, then generate one using `asyncapi new --example=tutorial.yml --no-tty` command and generate server using `asyncapi generate fromTemplate asyncapi.yaml @asyncapi/nodejs-template -o output -p server=mosquitto
> cd output && npm install`

Reviewer 3: ```suggestion
If you did not follow the previous tutorial and do not have an `asyncapi.yaml` file ready, then generate one using `asyncapi new --example=tutorial.yml --no-tty` command and generate server using below commands:
    ```
    asyncapi generate fromTemplate asyncapi.yaml @asyncapi/nodejs-template -o output -p server=mosquitto
    cd output && npm install
    ```
```

I suggest this as the way it is done correctly it is displayed as one command, it would fail when executed in terminal
